pkg/addonfactory/factory.go

```diff
@@ -0,0 +1,138 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+)
+
+const AddonDefaultInstallNamespace = "open-cluster-management-agent-addon"
+
+// AgentAddonMode defines the mode of agent addon deployment.
+// AgentAddonDefaultMode: only install manifests on the managed cluster with default AgentAddonOptions
+// including default CSRConfigurations,no csrSign,and auto-approve all the CSRs.
+// AgentAddonCustomMode: implements the agentAddon instance only with WithAgentAddon option.
+type AgentAddonMode string
+
+const (
+	AgentAddonDefaultMode AgentAddonMode = "AgentAddonDefaultMode"
+	AgentAddonCustomMode  AgentAddonMode = "AgentAddonCustomMode"
+)
+
+type AddonFactory interface {
+	Start(ctx context.Context)
+}
+
+// AddonOptions defines the functional option type for DefaultAddonFactory.
+type AddonOptions func(factory *addonFactory) *addonFactory
+
+type TemplateManifests func(config map[string]interface{}) ([]runtime.Object, error)
+
+type addonFactory struct {
+	kubeClient kubernetes.Interface
+	addonName  string
+	// staticManifests includes some static manifests deployed on the managed cluster,like CRDs, rbac, configMaps, etc.
+	staticManifests []runtime.Object
+	// templateManifests includes a given go template and config of manifests deployed on the managed cluster.
+	templateManifests TemplateManifests
+	// permissionManifests includes the rbac manifests deployed on the hub cluster.
+	permissionManifests []runtime.Object
+	registration        *agent.RegistrationOption
+	installStrategy     *agent.InstallStrategy
+	agentAddon          agent.AgentAddon
+	addonManager        addonmanager.AddonManager
+}
+
+func NewAddonFactoryWithOptions(agentAddonMode AgentAddonMode, kubeConfig *rest.Config, addonName string, options ...AddonOptions) AddonFactory {
+	kubeClient, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		klog.Exit("unable to create kube client: %v", err)
+	}
+
+	addonManager, _ := addonmanager.New(kubeConfig)
+
+	if len(addonName) == 0 {
+		addonName = fmt.Sprintf("addon-%s", utilrand.String(5))
+	}
+
+	factory := &addonFactory{
+		kubeClient:   kubeClient,
+		addonName:    addonName,
+		addonManager: addonManager,
+	}
+	// Apply all options
+	for _, opt := range options {
+		factory = opt(factory)
+	}
+
+	switch agentAddonMode {
+	case AgentAddonDefaultMode:
+		factory.agentAddon = newDefaultAgentAddon(factory)
+	case AgentAddonCustomMode:
+		factory.agentAddon = nil
+	default:
+		factory.agentAddon = newDefaultAgentAddon(factory)
+	}
+
+	return factory
+}
+
+func (a *addonFactory) Start(ctx context.Context) {
+	if a.agentAddon == nil {
+		klog.Exit(fmt.Errorf("invalid agentAddon defined"))
+	}
+
+	a.addonManager.AddAgent(a.agentAddon)
+	a.addonManager.Start(ctx)
+}
+
+func WithInstallStrategy(installAll bool, installNamespace string) AddonOptions {
```

Comment:
Why not `WithInstallStrategy(installAll bool, installNamespace string) *addonFactory`